examples/pebbleui/js/modules/PebbleUI.js

```diff
@@ -0,0 +1,561 @@
+      if (windowIndex >= 0) {
+        parent._windows.splice(windowIndex, 1);
+      }
+
+      parent.mark_dirty();
+      return this;
+    };
+
+    windowElement.add = function(el) {
+      var elementIndex = this._elements.indexOf(el);
+
+      // add a Z-Index if it's not already set
+      if (!el.z) el.z = this._currentZ++;
+
+      // Remove the elements from the stack if it exists
+      if (elementIndex >= 0) {
+        this._elements.splice(elementIndex, 1);
+      }
+
+      // Add the element to the list and sort
+      this._elements.push(el);
+      this._elements = this._elements.sort(function(a, b) {
+        return a.z - b.z;
+      });
+
+      parent.mark_dirty();
+      return this;
+    };
+
+    windowElement.remove = function(el) {
+      var elementIndex = this._elements.indexOf(el);
+      // Remove the elements from the stack if it exists
+      if (elementIndex >= 0) {
+        this._elements.splice(elementIndex, 1);
+      }
+
+      parent.mark_dirty();
+      return this;
+    };
+
+    windowElement._defaultBackHandler = function(event) {
+      // Remove the current window
+      this.hide();
+    };
+
+    windowElement.onUp = function(cb) {
+      this._upHandler = cb;
+      return this;
+    };
+
+    windowElement.onDown = function(cb) {
+      this._downHandler = cb;
+      return this;
+    };
+
+    windowElement.onSelect = function(cb) {
+      this._selectHandler = cb;
+      return this;
+    };
+
+    windowElement.onBack = function(cb) {
+      if (cb) this._backHandler = cb;
+      else this._backHandler = this._defaultBackHandler;
+      return this;
+    };
+
+    windowElement._invokeButtonCallbackByCode = function(code, event) {
+      switch (code) {
+        case 37:
+          if (this._backHandler) this._backHandler(event);
+          break;
+        case 38:
+          if (this._upHandler) this._upHandler(event);
+          break;
+        case 39:
+          if (this._selectHandler) this._selectHandler(event);
+          break;
+        case 40:
+          if (this._downHandler) this._downHandler(event);
+          break;
+        default:
+          console.log('Unknown keycode: ' + code);
+      }
+    };
+
+    // Set callbacks
+    windowElement._backHandler = options.onBack || windowElement._defaultBackHandler;
+    windowElement._selectHandler = options.onSelect;
+    windowElement._upHandler = options.onUp;
+    windowElement._downHandler = options.onDown;
+
+    windowElement._render = function(ctx, bounds) {
+      this._background._render(ctx, bounds);
+
+      // _render the remainder of the elements
+      this._elements.forEach(function(el) {
+        if (el._render) el._render(ctx, bounds);
+      });
+    };
+
+    return windowElement;
+  };
+
+  // A Card is not an Element
+  // But it is a Window (which is an Element)
+  this.Card = function(options) {
+    options = options || {};
+
+    var card = new parent.Window(options);
+
+    card._title = new parent.Text({
```

Comment:
with this concept where each UI element is a child of the `this` and not the logical owner it will be difficult to add animations of these structured elements. My previous comment "does parent implement `#mark_dirty()` was referring to this.